File: src/main/java/eu/happybe/openapi/party/Party.java
package eu.happybe.openapi.party;

import eu.happybe.openapi.mysql.QueryQueue;
import eu.happybe.openapi.mysql.query.AddPartyMemberQuery;
import eu.happybe.openapi.mysql.query.RemovePartyMemberQuery;
import eu.happybe.openapi.mysql.query.UpdateRowQuery;
import eu.happybe.openapi.servers.Server;
import eu.happybe.openapi.servers.ServerManager;
import io.gomint.entity.EntityPlayer;
import lombok.Getter;

import java.util.HashMap;
import java.util.Map;

public class Party {

    @Getter
    private boolean isOnline = true;
    @Getter
    private final EntityPlayer owner;

    @Getter
    private final Map<String, EntityPlayer> members = new HashMap<>();

    public Party(EntityPlayer owner) {
        this.owner = owner;
    }

    public void addMember(EntityPlayer player) {
        this.addMember(player, true);
    }

    public void addMember(EntityPlayer player, boolean updateInDatabase) {
        if(updateInDatabase) {
            QueryQueue.submitQuery(new AddPartyMemberQuery(this.getOwner().name(), player.name()));
        }

        this.getMembers().put(player.name(), player);
    }

    public void removeMember(EntityPlayer player) {
        this.removeMember(player, true);
    }

    public void removeMember(EntityPlayer player, boolean updateInDatabase) {
        if(updateInDatabase) {
            QueryQueue.submitQuery(new RemovePartyMemberQuery(this.getOwner().name(), player.name()));
        }

        this.getMembers().remove(player.name());
    }

    public boolean containsPlayer(EntityPlayer player) {
        return this.getMembers().containsKey(player.name());
    }

    public void broadcastMessage(String message) {
        for(EntityPlayer player : this.getAll().values()) {
            player.sendMessage(message);
        }
    }

    public void transfer(Server server) {
        QueryQueue.submitQuery(new UpdateRowQuery(new HashMap<String, Object> () {{
            this.put("CurrentServer", server.getServerName());
        }}, "Owner", this.getOwner().name(), "Parties"));

        this.isOnline = server.getServerName().equals(ServerManager.getCurrentServer().getServerName());

        server.transferPlayerHere(this.getOwner());
        for(EntityPlayer member : this.getMembers().values()) {
            server.transferPlayerHere(member);
        }

        if(!this.isOnline()) {
            PartyManager.removeParty(this);
        }
    }

    public Map<String, EntityPlayer> getAll() {
        Map<String, EntityPlayer> all = new HashMap<>(this.getMembers());
        all.put(this.getOwner().name(), this.getOwner());

        return all;
    }
}

File: src/main/java/eu/happybe/openapi/party/Party.kt
package eu.happybe.openapi.party

import cn.nukkit.Player
import eu.happybe.openapi.mysql.QueryQueue
import eu.happybe.openapi.mysql.query.AddPartyMemberQuery
import eu.happybe.openapi.mysql.query.RemovePartyMemberQuery
import eu.happybe.openapi.mysql.query.UpdateRowQuery
import eu.happybe.openapi.servers.Server
import eu.happybe.openapi.servers.ServerManager
import lombok.Getter
import java.util.*

class Party(@field:Getter private val owner: Player) {
    @Getter
    var isOnline = true

    @Getter
    private val members: Map<String, Player> = HashMap()
    @JvmOverloads
    fun addMember(player: Player, updateInDatabase: Boolean = true) {
        if (updateInDatabase) {
            QueryQueue.submitQuery(AddPartyMemberQuery(this.getOwner().getName(), player.name))
        }
        this.getMembers().put(player.name, player)
    }

    @JvmOverloads
    fun removeMember(player: Player, updateInDatabase: Boolean = true) {
        if (updateInDatabase) {
            QueryQueue.submitQuery(RemovePartyMemberQuery(this.getOwner().getName(), player.name))
        }
        this.getMembers().remove(player.name)
    }

    fun containsPlayer(player: Player): Boolean {
        return this.getMembers().containsKey(player.name)
    }

    fun broadcastMessage(message: String?) {
        for (player in all.values) {
            player.sendMessage(message)
        }
    }

    fun transfer(server: Server) {
        QueryQueue.submitQuery(UpdateRowQuery(object : HashMap<String?, Any?>() {
            init {
                this["CurrentServer"] = server.getServerName()
            }
        }, "Owner", this.getOwner().getName(), "Parties"))
        isOnline = server.getServerName() == ServerManager.getCurrentServer().getServerName()
        server.transferPlayerHere(this.getOwner())
        for (member in this.getMembers().values) {
            server.transferPlayerHere(member)
        }
        if (!this.isOnline()) {
            PartyManager.removeParty(this)
        }
    }

    val all: Map<String, Player>
        get() {
            val all: MutableMap<String, Player> = HashMap<String, Player>(this.getMembers())
            all[this.getOwner().getName()] = this.getOwner()
            return all
        }
}